crates/sui-types/src/passkey_authenticator.rs

```diff
@@ -0,0 +1,311 @@
+            return Err(SuiError::InvalidSignature {
+                error: "Invalid challenge".to_string(),
+            });
+        };
+
+        // Construct msg = authenticator_data || sha256(client_data_json).
+        let mut message = self.authenticator_data.clone();
+        let client_data_hash = Sha256::digest(self.client_data_json.as_bytes()).digest;
+        message.extend_from_slice(&client_data_hash);
+
+        // Check if author is derived from the public key.
+        if author != SuiAddress::from(&self.get_pk()?) {
+            return Err(SuiError::InvalidSignature {
+                error: "Invalid author".to_string(),
+            });
+        };
+
+        // Verify the signature against pk and message.
+        self.pk
+            .verify(&message, &self.signature)
+            .map_err(|_| SuiError::InvalidSignature {
+                error: "Fails to verify".to_string(),
+            })
+    }
+}
+
+impl ToFromBytes for PasskeyAuthenticator {
+    fn from_bytes(bytes: &[u8]) -> Result<Self, FastCryptoError> {
+        // The first byte matches the flag of PasskeyAuthenticator.
+        if bytes.first().ok_or(FastCryptoError::InvalidInput)?
+            != &SignatureScheme::PasskeyAuthenticator.flag()
+        {
+            return Err(FastCryptoError::InvalidInput);
+        }
+        let passkey: PasskeyAuthenticator =
+            bcs::from_bytes(&bytes[1..]).map_err(|_| FastCryptoError::InvalidSignature)?;
+        Ok(passkey)
+    }
+}
+
+impl AsRef<[u8]> for PasskeyAuthenticator {
+    fn as_ref(&self) -> &[u8] {
+        self.bytes
+            .get_or_try_init::<_, eyre::Report>(|| {
+                let as_bytes = bcs::to_bytes(self).expect("BCS serialization should not fail");
+                let mut bytes = Vec::with_capacity(1 + as_bytes.len());
+                bytes.push(SignatureScheme::PasskeyAuthenticator.flag());
+                bytes.extend_from_slice(as_bytes.as_slice());
+                Ok(bytes)
+            })
+            .expect("OnceCell invariant violated")
+    }
+}
+/// Compute the digest that the signature committed over as `intent || hash(tx_data)`, total
+/// of 3 + 32 = 35 bytes.
+pub fn to_signing_message<T: Serialize>(
+    intent_msg: &IntentMessage<T>,
+) -> [u8; INTENT_PREFIX_LENGTH + DefaultHash::OUTPUT_SIZE] {
+    let mut extended = [0; INTENT_PREFIX_LENGTH + DefaultHash::OUTPUT_SIZE];
+    extended[..INTENT_PREFIX_LENGTH].copy_from_slice(&intent_msg.intent.to_bytes());
+    extended[INTENT_PREFIX_LENGTH..].copy_from_slice(&to_signing_digest(intent_msg));
+    extended
+}
+
+/// Compute the hash(tx_data) from the intent msg.
+pub fn to_signing_digest<T: Serialize>(
+    intent_msg: &IntentMessage<T>,
+) -> [u8; DefaultHash::OUTPUT_SIZE] {
+    let mut hasher = DefaultHash::default();
+    bcs::serialize_into(&mut hasher, &intent_msg.value)
+        .expect("Message serialization should not fail");
+    hasher.finalize().digest
+}
```

Comment:
I thought we agreed on this being the Transaction Digest? I don't see the value in encoding the challenge with (Intent | non tx digest) over hash(Intent | non tx digest) in this case. The value that was being provided with this sending (Intent | digest) was that we could match the digest to the transaction digest that is well known.

Reply:
i misunderstood before - since it will end up diverging with the original signing digest i thought we were ok with signing_digest not tx_digest. 

while i can change to tx_digest, this introduces a non trivial change to the code that involves an additional trait everywhere that we wanted to avoid before

T: Serialize + Signable<DefaultHash>
instead of 
T: Serialize

for all authenticator trait implementers. is this desirable? is there an alternative way to avoid it?

Reply:
PersonalMessage also uses this trait but its not implemented inside the private trait module, it only implements Serialize

Reply:
After discussing it seems like we should leave this as-is given deviating from this would be thrashing against the status quo (which we've identified is not great but not something we can really worry about addressing right now)